File: ds_4_linkedlist/linkedlist_3_two_points/L141.cpp
//
// Created by LiHeng on 2020/4/19.
//

//判是否是环形链表

#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

//方法1 快慢指针法:注意快慢指针法只是利用只要有环，一定会相遇，但相遇节点并不一定是环的入口
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == NULL || head->next == NULL )
            return false ;
        ListNode* fast = head ;
        ListNode* slow = head ;
        while(fast->next != NULL && fast->next->next != NULL){
            fast = fast->next->next ;
            slow = slow->next ;
            if(fast == slow )
                return true ;
        }
        return false ;
    }
};


//采用哈希法，这里用集合就行了
#include <set>
class Solution1 {
public:
    bool hasCycle(ListNode *head){
        if(head == NULL || head->next == NULL ) //不能用&& 用||
            return false ;
        set<ListNode *>  p_set ;
        ListNode *p = head ;
        while(p){
            if(p_set.find(p) != p_set.end())
                return true ;
            p_set.insert(p) ;
            p = p->next ;
        }
        return false ;
    }
};

File: python-algorithm/ds_4_linkedlist/linkedlist_3_two_points/L141.py
#--coding:utf-8--
'''
@Time   : 2020/8/14
@Author : Heng Li
@Email  : liheng_lulia@163.com
'''


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if not head or not head.next:
            return False
        slow = fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                return True
        return False